File: babbleon/platform.py
"""
Platform detection and capability probes.

All platform-specific branching in the codebase should import from here,
not call os.uname() / sys.platform directly. This keeps hardware-specific
paths easy to audit and fake in tests.
"""

from __future__ import annotations

import os
import pathlib
import shutil
import sys


def current() -> str:
    """Return canonical platform identifier: 'linux', 'macos', 'windows', 'other'."""
    p = sys.platform
    if p.startswith("linux"):
        return "linux"
    if p == "darwin":
        return "macos"
    if p in ("win32", "cygwin"):
        return "windows"
    return "other"


def is_linux() -> bool:
    return current() == "linux"


def is_macos() -> bool:
    return current() == "macos"


# -- capability probes (lazy, cached) -----------------------------------------

_cache: dict[str, bool] = {}


def _probe(key: str, fn) -> bool:
    if key not in _cache:
        try:
            _cache[key] = bool(fn())
        except Exception:
            _cache[key] = False
    return _cache[key]


def has_unshare() -> bool:
    """True if CLONE_NEWNS unshare is likely to succeed (Linux + CAP_SYS_ADMIN)."""
    if not is_linux():
        return False
    return _probe("unshare", lambda: os.access("/proc/self/ns/mnt", os.R_OK))


def has_tpm2_tools() -> bool:
    return _probe("tpm2_tools", lambda: shutil.which("tpm2_getcap") is not None)


def has_proc_fs() -> bool:
    return _probe("proc", lambda: pathlib.Path("/proc/self/status").exists())


def kernel_version() -> tuple[int, int, int]:
    if not is_linux():
        return (0, 0, 0)
    r = os.uname().release.split("-")[0]
    parts = r.split(".")
    try:
        return tuple(int(p) for p in parts[:3])  # type: ignore[return-value]
    except ValueError:
        return (0, 0, 0)


def supports_landlock() -> bool:
    """Landlock requires kernel >= 5.13."""
    return is_linux() and kernel_version() >= (5, 13, 0)


def supports_ebpf_lsm() -> bool:
    """eBPF-LSM requires kernel >= 5.7 and lsm=...,bpf in cmdline."""
    if not is_linux() or kernel_version() < (5, 7, 0):
        return False
    return _probe("ebpf_lsm", _check_ebpf_lsm_boot_param)


def _check_ebpf_lsm_boot_param() -> bool:
    cmdline = pathlib.Path("/proc/cmdline")
    if not cmdline.exists():
        return False
    return "lsm=" in cmdline.read_text() and "bpf" in cmdline.read_text()


def clear_cache() -> None:
    """For tests: reset all cached probe results."""
    _cache.clear()

File: crates/babbleon/src/platform.rs
//! Single source of truth for platform detection.
//!
//! All platform branching in the codebase should call helpers here, not
//! `cfg!` directly. cfg-based compilation gating IS allowed in modules that
//! are entirely platform-specific (e.g. `enforcement::linux_ns`).

use std::path::Path;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Platform {
    Linux,
    MacOS,
    Windows,
    Other,
}

pub fn current() -> Platform {
    if cfg!(target_os = "linux") {
        Platform::Linux
    } else if cfg!(target_os = "macos") {
        Platform::MacOS
    } else if cfg!(target_os = "windows") {
        Platform::Windows
    } else {
        Platform::Other
    }
}

pub fn is_linux() -> bool {
    matches!(current(), Platform::Linux)
}

pub fn is_macos() -> bool {
    matches!(current(), Platform::MacOS)
}

/// True when /proc/self/ns/mnt is readable — necessary for unshare workflows.
pub fn has_unshare() -> bool {
    is_linux() && Path::new("/proc/self/ns/mnt").exists()
}

pub fn has_proc_fs() -> bool {
    Path::new("/proc/self/status").exists()
}

/// Best-effort parse of `uname -r` into (major, minor, patch).
pub fn kernel_version() -> (u32, u32, u32) {
    if !is_linux() {
        return (0, 0, 0);
    }
    let release = std::fs::read_to_string("/proc/sys/kernel/osrelease").unwrap_or_default();
    let head = release.split('-').next().unwrap_or("").trim();
    let mut parts = head.split('.').map(|p| p.parse::<u32>().unwrap_or(0));
    (
        parts.next().unwrap_or(0),
        parts.next().unwrap_or(0),
        parts.next().unwrap_or(0),
    )
}

/// Landlock requires Linux kernel >= 5.13.
pub fn supports_landlock() -> bool {
    is_linux() && kernel_version() >= (5, 13, 0)
}

/// eBPF-LSM requires kernel >= 5.7 AND lsm=...,bpf in /proc/cmdline.
pub fn supports_ebpf_lsm() -> bool {
    if !is_linux() || kernel_version() < (5, 7, 0) {
        return false;
    }
    let cmdline = std::fs::read_to_string("/proc/cmdline").unwrap_or_default();
    cmdline.contains("lsm=") && cmdline.contains("bpf")
}

/// Probe for tpm2-tools presence (used by the TPM backend's subprocess fallback).
pub fn has_tpm2_tools() -> bool {
    which::which_in("tpm2_getcap", std::env::var_os("PATH"), "/").is_ok()
}

/// Minimal `which` replacement to avoid pulling in the `which` crate.
mod which {
    use std::ffi::OsString;
    use std::path::PathBuf;

    pub enum WhichErr {
        NotFound,
    }

    pub fn which_in<P: AsRef<std::ffi::OsStr>>(
        name: &str,
        path_var: Option<OsString>,
        _cwd: P,
    ) -> Result<PathBuf, WhichErr> {
        let path = path_var.unwrap_or_default();
        for dir in std::env::split_paths(&path) {
            let candidate = dir.join(name);
            if candidate.exists() {
                return Ok(candidate);
            }
        }
        Err(WhichErr::NotFound)
    }
}
